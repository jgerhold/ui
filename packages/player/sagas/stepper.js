import { takeEvery, select, put } from 'redux-saga/effects'
import { STEP_FORWARD, STEP_BACKWARDS } from '@podlove/player-actions/types'
import { min, max } from 'ramda'
import { requestPlaytime } from '@podlove/player-actions/timepiece'

export const stepperSaga = ({ selectDuration, selectPlaytime }) =>
  function* saga() {
    yield takeEvery(STEP_FORWARD, stepForward, { selectDuration, selectPlaytime })
    yield takeEvery(STEP_BACKWARDS, stepBackward, { selectPlaytime })
  }

export function* stepForward({ selectDuration, selectPlaytime }) {
  const duration = yield select(selectDuration)
  const playtime = yield select(selectPlaytime)
  const targetPlaytime = min(playtime + 30 * 1000, duration)
  yield put(requestPlaytime(targetPlaytime))
}

export function* stepBackward({ selectPlaytime }) {
  const playtime = yield select(selectPlaytime)
  const targetPlaytime = max(playtime - 15 * 1000, 0)
  yield put(requestPlaytime(targetPlaytime))
}
