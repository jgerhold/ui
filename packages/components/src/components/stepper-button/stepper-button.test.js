import { mount } from '@vue/test-utils'
import { stepForward, stepBackwards } from '@podlove/player-actions/stepper'

import StepperButton from './StepperButton'

describe('StepperButton', () => {
  describe('component', () => {
    test('should be a Vue instance', async () => {
      const wrapper = await mount(StepperButton, {
        propsData: {
          type: 'forward'
        }
      })

      expect(wrapper.isVueInstance()).toBeTruthy()
    })
  })

  describe('props', () => {
    const types = ['forward', 'backwards']

    types.forEach(type => {
      test(`should render for rype ${type}`, async () => {
        const wrapper = await mount(StepperButton, {
          propsData: {
            type,
            color: '#fff'
          }
        })

        expect(wrapper.element).toMatchSnapshot()
      })
    })
  })

  describe('events', () => {
    test('should emit STEP_FORWARD on type forward click', async () => {
      const wrapper = await mount(StepperButton, {
        propsData: {
          type: 'forward'
        }
      })

      wrapper.trigger('click')

      expect(wrapper.emitted('click')).toEqual([[stepForward()]])
    })

    test('should emit STE_BACKWARD on type backward click', async () => {
      const wrapper = await mount(StepperButton, {
        propsData: {
          type: 'backwards'
        }
      })

      wrapper.trigger('click')

      expect(wrapper.emitted('click')).toEqual([[stepBackwards()]])
    })
  })
})
