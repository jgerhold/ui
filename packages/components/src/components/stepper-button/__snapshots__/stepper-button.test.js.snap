// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StepperButton props should render for rype backwards 1`] = `
<button
  class="stepper-button"
  id="stepper-button--backwards"
>
  <svg
    aria-hidden="true"
    background="currentColor"
    height="48.589999999999996"
    version="1.1"
    viewBox="0 0 37 32"
    width="43"
    xml:space="preserve"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <g>
      <g
        transform="translate(2.000000, -9.000000)"
      >
        <g
          transform="translate(0.000000, 10.000000)"
        >
          <path
            d="M32.3,14.9c0-7.3-6-13.3-13.3-13.3C12.5,1.7,6.8,6.5,5.9,13l-0.1,0.9H2.7l4.1,5.5l4.1-5.5H7.8l0.2-1.2
          c1.1-5.2,5.7-9,11-9c6.2,0,11.2,5,11.2,11.2c0,5.8-4.5,10.6-10.2,11.2v2.1C26.9,27.7,32.3,21.9,32.3,14.9z"
            fill="#fff"
          />
           
          <g>
            <path
              d="M17.2,19.5h-1.4v-7.2H14v-1l3.2-0.3V19.5z M18.7,15.7l0.5-4.7h4.5v1.2h-3.3l-0.3,2.1
            c0.2-0.1,0.4-0.2,0.6-0.3c0.2-0.1,0.5-0.1,0.7-0.1c0.8,0,1.4,0.2,1.9,0.8c0.5,0.5,0.7,1.2,0.7,2.1c0,0.8-0.2,1.5-0.7,2.1
            c-0.5,0.5-1.2,0.8-2.1,0.8c-0.8,0-1.4-0.2-1.9-0.6c-0.5-0.4-0.8-1-0.7-1.8l0,0l1.3,0c0,0.4,0.1,0.7,0.4,1
            c0.2,0.2,0.6,0.3,0.9,0.3c0.4,0,0.8-0.2,1-0.5c0.2-0.3,0.3-0.7,0.3-1.2c0-0.5-0.1-0.9-0.4-1.3c-0.2-0.3-0.6-0.5-1-0.5
            c-0.4,0-0.7,0.1-0.9,0.2c-0.2,0.1-0.3,0.3-0.4,0.5L18.7,15.7z"
              fill="#fff"
            />
          </g>
        </g>
      </g>
    </g>
  </svg>
   
</button>
`;

exports[`StepperButton props should render for rype forward 1`] = `
<button
  class="stepper-button"
  id="stepper-button--forward"
>
  <svg
    aria-hidden="true"
    background="currentColor"
    height="48.589999999999996"
    version="1.1"
    viewBox="0 0 36 31"
    width="43"
    xml:space="preserve"
    xmlns="http://www.w3.org/2000/svg"
    xmlns:xlink="http://www.w3.org/1999/xlink"
  >
    <g>
      <g
        transform="translate(-143.000000, -9.000000)"
      >
        <g
          transform="translate(144.000000, 10.000000)"
        >
          <path
            d="M30.6,13.2h-3.1l-0.1-0.9C26.5,5.9,20.8,1,14.3,1C7,1,1,7,1,14.3c0,7,5.4,12.7,12.3,13.2v-2.1
          C7.6,24.9,3.1,20.1,3.1,14.3c0-6.2,5-11.2,11.2-11.2c5.3,0,9.9,3.8,11,9l0.2,1.2h-3.1l4.1,5.5L30.6,13.2z"
            fill="#fff"
          />
           
          <g>
            <path
              d="M10.7,11c-1,0-1.8,0.3-2.5,1L9,12.9c0.5-0.5,1-0.7,1.6-0.7c0.7,0,1.2,0.4,1.2,1.1c0,0.8-0.5,1.2-1.3,1.2
            h-0.4l-0.2,1.1h0.6c0.9,0,1.5,0.3,1.5,1.3c0,0.8-0.6,1.4-1.4,1.4c-0.6,0-1.2-0.2-1.7-0.8L8,18.2c0.6,0.8,1.5,1.2,2.6,1.2
            c1.8,0,2.9-1.1,2.9-2.5c0-1.2-0.8-1.9-1.8-2c0.9-0.3,1.6-0.9,1.6-1.9C13.3,11.9,12.4,11,10.7,11z M17.4,11
            c1.9,0,2.9,1.5,2.9,4.2c0,2.7-1,4.2-2.9,4.2c-1.9,0-2.9-1.5-2.9-4.2C14.6,12.5,15.5,11,17.4,11z M17.4,12.2
            c-0.9,0-1.3,0.8-1.3,3c0,2.3,0.4,3,1.3,3c0.9,0,1.3-0.8,1.3-3C18.8,12.9,18.3,12.2,17.4,12.2z"
              fill="#fff"
            />
          </g>
        </g>
      </g>
    </g>
  </svg>
   
</button>
`;
